src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3029,25 +3083,58 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
         public override BoundNode? VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
         {
             var localFunc = node.Symbol;
+
+            // Usages state is created when we visit the function's call site or its body.
+            // In the first pass of the nullable walker, existence of usages here means that a call site has been visited.
+            // In subsequent nullable walker passes, the starting state is preserved from previous passes.
+            // In any case, existence of usages means that we have a good starting state.
+            var hasGoodStartingState = HasLocalFuncUsagesCreated(localFunc);
+
             var localFunctionState = GetOrCreateLocalFuncUsages(localFunc);
 
-            // The starting state (`state`) is the top state ("maybe null").
+            // The state for the function's body analysis starts as the top state ("maybe null").
             var state = TopState();
 
-            // Captured variables are joined with the state
-            // at all the local function use sites (`localFunctionState.StartingState`)
-            // which starts as the bottom state ("not null").
-            var startingState = localFunctionState.StartingState;
-            startingState.ForEach(
-                (slot, variables) =>
-                {
-                    var symbol = variables[variables.RootSlot(slot)].Symbol;
-                    if (Symbol.IsCaptured(symbol, localFunc))
+            if (!hasGoodStartingState)
+            {
+                // For unreachable local functions, top-level captured variables are set to "not null",
+                // ignoring nested slots to avoid depending on slot allocation order
+                // (e.g., whether we have seen a class field or not already).
+                state.Normalize(this, _variables);
+                state.ForEach(
+                    (slot, variables) =>
                     {
-                        SetState(ref state, slot, GetState(ref startingState, slot));
-                    }
-                },
-                _variables);
+                        if (variables[slot].ContainingSlot == 0) // top-level variables only
+                        {
+                            var symbol = variables[variables.RootSlot(slot)].Symbol;
```

Comment:
>variables.RootSlot [](http://example.com/codeflow?start=51&length=18)

It doesn't look like this call is necessary. Regardless whether we keep the enclosing `if` or not.